performance fixes

arnetheduck opened the pull request:
* don't mark tree cache as dirty on read-only List accesses
* store only blob in memory for keys and signatures, parse blob lazily
* compare public keys by blob instead of parsing / converting to raw
* compare Eth2Digest using non-constant-time comparison
* avoid some unnecessary validator copying

This branch will in particular speed up deposit processing which has
been slowing down block replay.

Pre (mainnet, 1400 blocks):

```
All time are ms
     Average,       StdDev,          Min,          Max,      Samples,         Test
Validation is turned off meaning that no BLS operations are performed
    3450.269,        0.000,     3450.269,     3450.269,            1, Initialize DB
       0.417,        0.822,        0.036,       21.098,         1400, Load block from database
      16.521,        0.000,       16.521,       16.521,            1, Load state from database
      27.906,       50.846,        8.104,     1507.633,         1350, Apply block
      52.617,       37.029,       20.640,      135.938,           50, Apply epoch block
```

Post:

```
    3502.715,        0.000,     3502.715,     3502.715,            1, Initialize DB
       0.080,        0.560,        0.035,       21.015,         1400, Load block from database
      17.595,        0.000,       17.595,       17.595,            1, Load state from database
      15.706,       11.028,        8.300,      107.537,         1350, Apply block
      33.217,       12.622,       17.331,       60.580,           50, Apply epoch block
```

arnetheduck committed:
performance fixes

* don't mark tree cache as dirty on read-only List accesses
* store only blob in memory for keys and signatures, parse blob lazily
* compare public keys by blob instead of parsing / converting to raw
* compare Eth2Digest using non-constant-time comparison
* avoid some unnecessary validator copying

This branch will in particular speed up deposit processing which has
been slowing down block replay.

Pre (mainnet, 1600 blocks):

```
All time are ms
     Average,       StdDev,          Min,          Max,      Samples,         Test
Validation is turned off meaning that no BLS operations are performed
    3450.269,        0.000,     3450.269,     3450.269,            1, Initialize DB
       0.417,        0.822,        0.036,       21.098,         1400, Load block from database
      16.521,        0.000,       16.521,       16.521,            1, Load state from database
      27.906,       50.846,        8.104,     1507.633,         1350, Apply block
      52.617,       37.029,       20.640,      135.938,           50, Apply epoch block
```

Post:

```
    3502.715,        0.000,     3502.715,     3502.715,            1, Initialize DB
       0.080,        0.560,        0.035,       21.015,         1400, Load block from database
      17.595,        0.000,       17.595,       17.595,            1, Load state from database
      15.706,       11.028,        8.300,      107.537,         1350, Apply block
      33.217,       12.622,       17.331,       60.580,           50, Apply epoch block
```

tersec commented on beacon_chain/spec/beaconstate.nim:
```diff
@@ -91,31 +92,36 @@ proc process_deposit*(preset: RuntimePreset,
+  # build the table and use it for lookups than to scan it like this.
+  # Once we have a reusable, long-lived cache, this should be revisited
+  for i in 0..<state.validators.len():
+    if state.validators.asSeq()[i].pubkey == pubkey:
```
One reason to have done the non-pure-byte-comparison was that the bit patterns in memory aren't necessary identical if they're not in the same serialized/deserialized state, and the type system wasn't verifying that either way.

arnetheduck replied:
well, the byte comparisons were already being done *sometimes* (depending on whether `initPubKey` had been called or not) - from what I know, at least the valid BLS points each have unique representations make byte comparisons ok. the invalid ones, well, they don't significantly matter as far as equality is concerned - we won't find them where equality is required.

tersec commented on beacon_chain/spec/beaconstate.nim:
```diff
@@ -91,31 +92,36 @@ proc process_deposit*(preset: RuntimePreset,
 
   let
     pubkey = deposit.data.pubkey
-    pubkey_inited = pubkey.initPubKey
```
`initPubKey` should (as in, it was designed to) have a fast path that if it was already run on that particular key, it'll just memoize/use its cache. So its failing to do so here points to a deeper issue.

arnetheduck replied:
well, yes and no - what happens is that iintializing the pubKey in `state.validators` means a lookup at least which, compared to just comparing bytes, is slow (because in state.validators, the key was always not memoized in the instance, because when replaying, the state comes from disk)

zah commented on beacon_chain/spec/crypto.nim:
```diff
@@ -83,52 +75,54 @@ export AggregateSignature
+
+proc load*(v: ValidatorPubKey): Option[blscurve.PublicKey] =
+  ## Parse public key blob - this may fail
+  var cache {.threadvar.}: Table[typeof(v.blob), blscurve.PublicKey]
```
Can a remote attacker add entries to this cache somehow?

arnetheduck replied:
thought about that - they could potentially create blocks with deposits in them that have bogus public keys - seems like a long shot though as only blocks that are signed by an existing validator make it so far that pubkeys are processed - the alternative would be to limit this cache to trusted public keys loaded from states

arnetheduck replied:
renamed and documented

zah commented on beacon_chain/spec/digest.nim:
```diff
@@ -99,10 +99,12 @@ template withEth2Hash*(body: untyped): Eth2Digest =
-  # We just slice the first 4 or 8 bytes of the block hash
-  # depending of if we are on a 32 or 64-bit platform
-  result = cast[ptr Hash](unsafeAddr x)[]
+template `==`*(a, b: Eth2Digest): bool =
```
Since `Eth2Digest` is not a distinct type, this is not quite legal to do in Nim. The overload will affect other usages of `MDigest[256]` as well.

arnetheduck replied:
mm.. it's a bit of an annoyance, because there's no way to specify a custom `==` for `Table` for example, like in other languages - any ideas? could make eth2digest distinct I guess, and re-expose `data`.

zah commented on beacon_chain/spec/state_transition_epoch.nim:
```diff
@@ -548,16 +548,17 @@ func process_final_updates*(state: var BeaconState) {.nbench.}=
-  for index, validator in state.validators:
-    let balance = state.balances[index]
+  for index in 0..<state.validators.len:
+    let balance = state.balances.asSeq()[index]
```
How come operator `[]` is not defined over `SszList`? The dozens of calls to `asSeq` shouldn't be necessary.

arnetheduck replied:
it is, buy nim prefers the `var` overload which marks the entry as dirty.

arnetheduck replied:
ie the var overload of `[]` will be chosen when the seq is `var`, or by extension, when `BeaconState` is `var` - I don't know that there's any way to make it choose the more restrictive version - can't cast a `var` to a `let` like you can cast to `const` in `c++` afaik?

arnetheduck replied:
documented

mratsim commented on beacon_chain/eth2_json_rpc_serialization.nim:
```diff
@@ -15,7 +15,10 @@ proc toJsonHex(data: openArray[byte]): string =
+  var tmp = ValidatorPubKey.fromHex(n.getStr()).tryGet()
+  if not tmp.load().isSome():
+    raise (ref ValueError)(msg: "Invalid public BLS key")
+  result = tmp
```
We should have `raises: ValueError` in the signature

arnetheduck replied:
hm, that's probably reasonable, but also part of a separate refactoring commit - ie we raise resulterror here and a bunch of other things probably.

mratsim commented on beacon_chain/spec/state_transition_block.nim:
```diff
@@ -149,8 +149,8 @@ proc check_proposer_slashing*(
   # Verify the proposer is slashable
-  let proposer = state.validators[header_1.proposer_index]
-  if not is_slashable_validator(proposer, get_current_epoch(state)):
+  let proposer = unsafeAddr state.validators.asSeq()[header_1.proposer_index]
```
There is
```
import std/decls

let proposer {.byaddr.} = state.validators.asSeq()[header_1.proposer_index]
```

That is available for this pattern.

We can also reimplement a similar thing to have `{.byunsafeaddr.}` in stew as it seems like we are starting to use that quite often for optimization.

arnetheduck replied:
hmm, I'm a bit wary of these very exotic annotations that have low penetration in the language and are mostly workarounds - ie I don't quite see the difference between typing out `{.unsafeaddr.}` vs `unsafeAddr` - it's equally ugly, the shared problem being that I have to reason about implementation details and not express intent - except now, I also have to look it up in some weird module and learn yet another magic way of saying essentially the same thing - is there any notable difference?

mratsim replied:
It auto-dereferences so there is no ugly `[]` everywhere that clutter the code: https://github.com/nim-lang/Nim/blob/version-1-2/lib/std/decls.nim#L4-L19

arnetheduck replied:
hm, yeah, they extra dereferences are a feature to me, remind me that this is not a reference really but a pointer, as far as language rules are concerned - if they were to cross a closure boundary for example, they'd go stale - in C++ you have to go out of your way for that to happen but in nim, it's hidden because it captures everything automatically.

arnetheduck replied:
sticking with `unsafeAddr` to maintain a consistent style - `std/decls` looks like a footgun to me, but if we start using it consistently across the whole codebase, it could be used here as well (I'd recommend against though)

mratsim reviewed (commented):
We might want to rename `load` to `loadWithCaching` to make clear what's going on and also detect more easily places where it's unsafe to have cache because the keys didn't pass some necessary validation. (@zah's remark on exposing that cache to potential malicious parties).

In terms of merging, this will conflict the the Batch BLS changes as it transforms the whole crypto module, though it does simplify conversion because now TrustedSig can be a distinct type to ValidatorSig.

arnetheduck commented:
> In terms of merging, this will conflict the the Batch BLS changes

yeah, I'm unsure which changes should go in first - this branch is something of a critical bugfix to the extremely slow deposit processing - it crossed my mind that perhaps this branch should be cherry-picked into a release asap, not sure about the other one that also changes the bls submodule and the impl therein.

mratsim commented:
It's better to do the urgent before the important, I can rebase the BLS branch afterwards.

arnetheduck commented:
I just remembered how to put two things in a single graph with matplotlib :) last 10k slots of mainnet processing, orange is after the perf fixes are applied - it's quite the difference - though the signature check still makes its mark.

![prepost](https://user-images.githubusercontent.com/1382986/105572684-f8f14a00-5d58-11eb-92dd-064c6f560255.png)

arnetheduck committed:
more perf fixes

* load EpochRef cache into StateCache more aggressively
* point out security concern with public key cache
* reuse proposer index from state when processing block
* avoid genericAssign in a few more places
* don't parse key when signature is unparseable
* fix `==` overload for Eth2Digest
* preallocate validator list when getting active validators
* speed up proposer index calculation a little bit
* reuse cache when replaying blocks in ncli_db
* avoid a few more copying loops

```
     Average,       StdDev,          Min,          Max,      Samples,         Test
Validation is turned off meaning that no BLS operations are performed
    3279.158,        0.000,     3279.158,     3279.158,            1, Initialize DB
       0.072,        0.357,        0.035,       13.400,         1400, Load block from database
      17.295,        0.000,       17.295,       17.295,            1, Load state from database
       5.918,        9.896,        0.198,       98.028,         1350, Apply block
      15.888,       10.951,        7.902,       39.535,           50, Apply epoch block
       0.000,        0.000,        0.000,        0.000,            0, Database block store
```

arnetheduck commented:
```
     Average,       StdDev,          Min,          Max,      Samples,         Test
Validation is turned off meaning that no BLS operations are performed
    3279.158,        0.000,     3279.158,     3279.158,            1, Initialize DB
       0.072,        0.357,        0.035,       13.400,         1400, Load block from database
      17.295,        0.000,       17.295,       17.295,            1, Load state from database
       5.918,        9.896,        0.198,       98.028,         1350, Apply block
      15.888,       10.951,        7.902,       39.535,           50, Apply epoch block
       0.000,        0.000,        0.000,        0.000,            0, Database block store
```

this is a range that has a lot of deposits - on non-deposit ranges, applying a block is now around 1.5 ms, meaning replaying the whole mainnet chain takes about 1000s, including reading from db.

arnetheduck commented:
Remaining hotspots from running ncli_db bench on the 360k mainnet blocks

![Screenshot](https://user-images.githubusercontent.com/1382986/105643493-75d30f80-5e90-11eb-924f-1f4e90ba6b26.png)

arnetheduck committed:
clear full balance cache before processing rewards and penalties

```
All time are ms
     Average,       StdDev,          Min,          Max,      Samples,         Test
Validation is turned off meaning that no BLS operations are performed
    3947.901,        0.000,     3947.901,     3947.901,            1, Initialize DB
       0.124,        0.506,        0.026,      202.370,       363345, Load block from database
      97.614,        0.000,       97.614,       97.614,            1, Load state from database
       0.186,        0.188,        0.012,       99.561,       357262, Advance slot, non-epoch
      14.161,        5.966,        1.099,      395.511,        11524, Advance slot, epoch
       1.372,        4.170,        0.017,      276.401,       363345, Apply block, no slot processing
       0.000,        0.000,        0.000,        0.000,            0, Database block store
```

mratsim commented:
We can't do much on blst_sha256_block_data_order, it's using SSE3 (256-bit wide) and SHA256 is 256-bit wide so there is no simple AVX shortcut to process more data with the same number of instructions.
On Ryzen CPUs and Ice-lake we can make sure that we activate the SHA-NI extensions via some gcc flags but that's all.

Regarding `shuffle_list` it calls SHA256 in a loop in 2 places line 47 and 93 so we need to know if the `shuffle_list` time in your VTune (?) panel excludes blst_sha256_block_data_order or not https://github.com/status-im/nimbus-eth2/blob/87955f2d37c7b7e8b9f8ae1c47c9d354c59a1c34/beacon_chain/spec/validator.nim#L26-L93

mratsim commented:
Regarding bottlenecks, ncli_db does not bench the quarantine -> chainDAG transition btw so there might be other hotspot there, signature verification is an obvious one but we might have other quadratic algorithms in there.

Similarly I think it does not test attestation aggregation speed.

arnetheduck commented:
> We can't do much on blst_sha256_block_data_order,

one of the major improvements in this branch comes from not calling that function at all :) ie the `asSeq` stuff basically avoids dirtying the cache which reduces the number of calls quite drastically.

> `shuffle_list`

a lot of time is spent swapping bytes apparently - I suspect it's because this is not a very cache-friendly operation. some time is
 spent on counters which I suspect relates to bounds checking for integers. I don't know how accurate this is though - most of this branch is about avoiding operations like parsing keys and invalidating caches rather than micro-optimizations of the operations themselves.

![Screenshot from 2021-01-25 11-01-33](https://user-images.githubusercontent.com/1382986/105690944-e9126b00-5efc-11eb-88b8-357a9ae37ceb.png)

> vtune

vtune indeed - expanding the `blst_hash` function, most call stacks involve tree hashing, not `shuffle_list` - that said, it looks like the `shuffle_list` time doesn't include hashing.

> Regarding bottlenecks

Indeed, there's a lot more going on, but `ncli_db` fairly accurately models `EpochRef` cache misses and the subsequent `updateStateData` replays that happen to produce a new state/EpochRef which is where some nodes were struggling - ie it's when replaying many deposit blocks together that the worst slowdowns happen.

tersec commented:
> Regarding bottlenecks, ncli_db does not bench the quarantine -> chainDAG transition btw so there might be other hotspot there, signature verification is an obvious one

https://github.com/status-im/nimbus-eth2/pull/2254 helps for signature verificationk, by simply verifying fewer signatures.

arnetheduck commented:
the last commit includes a fix to call `clearCaches` less often:

![Screenshot from 2021-01-25 11-09-17](https://user-images.githubusercontent.com/1382986/105691662-d482a280-5efd-11eb-8a3d-c149a6170c43.png)

arnetheduck commented:
one can see a corresponding increase in hash calls because it's more blunt, but still looks like an over-all win.

arnetheduck commented:
more detailed timings when reusing the state cache:

```
     Average,       StdDev,          Min,          Max,      Samples,         Test
Validation is turned off meaning that no BLS operations are performed
    3947.901,        0.000,     3947.901,     3947.901,            1, Initialize DB
       0.124,        0.506,        0.026,      202.370,       363345, Load block from database
      97.614,        0.000,       97.614,       97.614,            1, Load state from database
       0.186,        0.188,        0.012,       99.561,       357262, Advance slot, non-epoch
      14.161,        5.966,        1.099,      395.511,        11524, Advance slot, epoch
       1.372,        4.170,        0.017,      276.401,       363345, Apply block, no slot processing
```
